### PR TITLE
Sharpen up descendant ordering terminology

### PR DESCRIPTION
**Pull request by glyn**
This is probably a bit too wordy, but is a starting point for a sharper definition of the nodelist orders which can be produced by the descendant selector.

Ref: https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/issues/215

### CODE REVIEW

File: draft-ietf-jsonpath-base.md [hunk 1]
```diff
@@ -1068,22 +1068,22 @@ Note that there is no "bald" descendant selector `..`.
...
+A `descendant-selector` selects certain descendants of a node:
 
 * the `..<name>` form (and the `..[<index>]` form where `<index>` is a `quoted-member-name`) selects those descendants that are member values of an object with the given member name.
 * the `..[<index>]` form, where `<index>` is an `element-index`, selects those descendants that are array elements with the given index.
 * the `..[<slice-index>]` form selects those descendants that are array elements selected by the given slice.
 * the `..[<filter>]` form selects those descendants that are array elements or object values selected by the given filter.
 * the `..[*]` and `..*` forms select all the descendants.
 
-The resultant nodelist is ordered as if:
+An _array preorder_ of the descendants of a node is a sequence of all the descendants in which:
```
Comment (cabo) [inline draft-ietf-jsonpath-base.md:1079]: Hmm, the "preorder" part is universal and not just for arrays.  So qualifying "preorder" with "array" doesn't strike me as helpful.  Simply calling it "preorder" is fine, but I wanted to add an adjective that indicates that this conventional order is one of many possible (JSON objects), which the implementation is free to choose, and that it is the order we have chosen for JSONPath.

Comment (glyn) [reply draft-ietf-jsonpath-base.md:1079]: Yeah, I was trying to capture an additional constraint on the order: it is both a preorder (parents before children) and arrays are traversed in array order (not implied by preorder).

So I tried to introduce a new term "array preorder" which is a special kind of preorder.

I'm not sure about an adjective conveying the lack of ordering of objects. Even if we found a suitable adjective, we'd still need to build in the array condition, ideally into the term, somehow.

Comment (cabo) [reply draft-ietf-jsonpath-base.md:1079]: After some more thinking, I don't get a better result than "array-ordered pre-order".  Maybe  bit redundant and wordy, but we are not using this term all over the place.

Comment (glyn) [reply draft-ietf-jsonpath-base.md:1079]: That's certainly explicit, but the ordered/order part doesn't sound good. Do any of the following grab you?

* array-based preorder
* arraywise preorder
* array-respecting preorder
* arrayful preorder
* linear preorder
* vector preorder

Comment (cabo) [reply draft-ietf-jsonpath-base.md:1079]: Surprisingly hard.

* array-sequenced pre-order
* array-guided pre-order

File: draft-ietf-jsonpath-base.md [hunk 2]
```diff
@@ -1068,22 +1068,22 @@ Note that there is no "bald" descendant selector `..`.
...
 
-* nodes are visited before their children, and
-* nodes of an array are visited in array order.
+* nodes of an array appear in array order, and
+* nodes appear before their children.
```
Comment (cabo) [inline draft-ietf-jsonpath-base.md:1082]: hmm, this doesn't actually say depth first.

..* on
[1, 2, 3, [4, 5], 6]

could be shown as [1, 2, 3, [4, 5], 6, 4, 5] and satisfy the two rules.

File: draft-ietf-jsonpath-base.md [hunk 3]
```diff
@@ -1068,22 +1068,22 @@ Note that there is no "bald" descendant selector `..`.
...
 * the `..[<slice-index>]` form selects those descendants that are array elements selected by the given slice.
 * the `..[<filter>]` form selects those descendants that are array elements or object values selected by the given filter.
 * the `..[*]` and `..*` forms select all the descendants.
 
-The resultant nodelist is ordered as if:
+An _array preorder_ of the descendants of a node is a sequence of all the descendants in which:
 
-* nodes are visited before their children, and
-* nodes of an array are visited in array order.
+* nodes of an array appear in array order, and
```
Comment (cabo) [inline draft-ietf-jsonpath-base.md:1081]:
```suggestion
* nodes of an array, immediately followed by their descendants appear in array order, and
```

Need to say the same for objects and their descendants.

File: draft-ietf-jsonpath-base.md (hunk 1 above)
Comment (glyn) [reply draft-ietf-jsonpath-base.md:1079]: "array-sequenced preorder" seems to be the best term so far. (I think preorder is usually unhyphenated. See, for example, [wikipedia](https://en.wikipedia.org/wiki/Preorder).)

File: draft-ietf-jsonpath-base.md (hunk 2 above)
Comment (glyn) [reply draft-ietf-jsonpath-base.md:1082]: Good point. I'll address that shortly.

**Review by sthagen — APPROVED** [review]
In case `a` and `an` Indicate `any` or `for every` this change looks good to me.

File: draft-ietf-jsonpath-base.md [hunk 4]
```diff
@@ -1063,27 +1063,27 @@ descendant-selector = ".." ( dot-member-name      /  ; ..<name>
                            )
 ~~~~
 
-Note that there is no "bald" descendant selector `..`.
+Note that there is no "bald" descendant selector (`..` on its own).
```
Comment (cabo) [inline draft-ietf-jsonpath-base.md:1066]: I'm not sure we need to introduce that term (even though, informally, I find it amusing).

```suggestion
Note that `..` on its own is not a valid selector.
```

**Review by cabo — APPROVED** [review]
(Optional comment.)

File: draft-ietf-jsonpath-base.md [hunk 5]
```diff
@@ -1063,27 +1063,27 @@ descendant-selector = ".." ( dot-member-name      /  ; ..<name>
...
 
 #### Semantics
 {: unnumbered}
 
-The `descendant-selector` selects certain descendants of a node:
+A `descendant-selector` selects certain descendants of a node:
```
Comment (sthagen) [inline draft-ietf-jsonpath-base.md:1071]: Does `A` Indicate `any` or `for every`?

File: draft-ietf-jsonpath-base.md [hunk 6]
```diff
@@ -1063,27 +1063,27 @@ descendant-selector = ".." ( dot-member-name      /  ; ..<name>
...
 
 * the `..<name>` form (and the `..[<index>]` form where `<index>` is a `quoted-member-name`) selects those descendants that are member values of an object with the given member name.
 * the `..[<index>]` form, where `<index>` is an `element-index`, selects those descendants that are array elements with the given index.
 * the `..[<slice-index>]` form selects those descendants that are array elements selected by the given slice.
 * the `..[<filter>]` form selects those descendants that are array elements or object values selected by the given filter.
 * the `..[*]` and `..*` forms select all the descendants.
 
-The resultant nodelist is ordered as if:
+An _array-sequenced preorder_ of the descendants of a node is a sequence of all the descendants in which:
```
Comment (sthagen) [inline draft-ietf-jsonpath-base.md:1079]: Does  `An` Indicate `any` or `for every`?

File: draft-ietf-jsonpath-base.md (hunk 5 above)
Comment (glyn) [reply draft-ietf-jsonpath-base.md:1071]: I changed "The" to "A" because there are multiple descendant selectors. It carries the connotation of "any".

File: draft-ietf-jsonpath-base.md (hunk 4 above)
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1066]: Not a good suggestion?

Comment (glyn) [reply draft-ietf-jsonpath-base.md:1066]: Whoops! Yes, it was a good suggestion, thanks. Since it's small, I'll piggy-back it on another PR.

Comment (glyn) [reply draft-ietf-jsonpath-base.md:1066]: See https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/pull/214/commits/c7c245a95302c5e792d34e0ba4c6c9de10e431b2
